src/builtins.ts

```diff
@@ -1171,14 +1171,22 @@ function builtin_rotl(ctx: BuiltinContext): ExpressionRef {
   if (type.isValue) {
     let arg1 = compiler.compileExpression(operands[1], type, Constraints.CONV_IMPLICIT);
     switch (type.kind) {
+      case TypeKind.BOOL: return arg0;
       case TypeKind.I8:
       case TypeKind.I16:
       case TypeKind.U8:
-      case TypeKind.U16:
-      case TypeKind.BOOL: {
-        return compiler.ensureSmallIntegerWrap(
-          module.binary(BinaryOp.RotlI32, arg0, arg1),
-          type
+      case TypeKind.U16: {
+        // (value << (shift & mask)) | (value >>> ((0 - shift) & mask))
+        let mask   = module.i32(type.size - 1);
```

Comment:
The mask expression is used twice below, but according Binaryen there is a risk when doing that, because optimizations would modify both places. Suggesting to inline all of `mask`, `shiftL` and `shiftR` into the returned binary as it also indicates that these are one-time uses.